src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -154,20 +228,101 @@ private void registerMBean()
 
     public void close()
     {
-        updateSchedular.cancel(false);
-        resetSchedular.cancel(false);
+        if (updateScheduler != null)
+            updateScheduler.cancel(false);
+        if (latencyProbeScheduler != null)
+            latencyProbeScheduler.cancel(false);
 
         MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
         try
         {
-            mbs.unregisterMBean(new ObjectName(mbeanName));
+            if (mbeanRegistered)
+                mbs.unregisterMBean(new ObjectName(mbeanName));
         }
         catch (Exception e)
         {
             throw new RuntimeException(e);
         }
     }
 
+    /**
+     * Determines if latency probes need to be sent, and potentially sends a single latency probe per invocation
+     */
+    protected void maybeSendLatencyProbe()
+    {
+        long currentTimeNS = System.nanoTime();
+        markNextAllowedProbeGenerationTime(currentTimeNS);
+
+        Optional<InetAddressAndPort> needsProbe = latencyProbeNeeded(currentTimeNS);
+        needsProbe.ifPresent(this::sendPingMessageToPeer);
+    }
+
+    /**
+     * Determines which peers need latency at a particular time. Note that this takes currentTimeNS for testability
+     * of this code path.
+     * @param currentTimeNS The current time to evaluate. Used mostly for testing.
+     * @return An Optional that if present contains a host to probe.
+     */
+    @VisibleForTesting
+    Optional<InetAddressAndPort> latencyProbeNeeded(long currentTimeNS) {
+        if (currentProbePosition >= latencyProbeSequence.size() && (currentTimeNS > nextProbeGenerationTime))
+        {
+            nextProbeGenerationTime = nextAllowedProbeGenerationTime;
+            latencyProbeSequence.clear();
+
+            // Delegate to the subclass to actually figure out what the probe sequence should be
+            updateLatencyProbeSequence(latencyProbeSequence);
+
+            if (latencyProbeSequence.size() > 0)
+                Collections.shuffle(latencyProbeSequence);
+
+            currentProbePosition = 0;
+        }
+
+        if (currentProbePosition < latencyProbeSequence.size())
+        {
+            try
+            {
+                return Optional.of(latencyProbeSequence.get(currentProbePosition++));
+            }
+            catch (IndexOutOfBoundsException ignored) {}
+        }
+
+        return Optional.empty();
+    }
+
+    private void sendPingMessageToPeer(InetAddressAndPort to)
+    {
+        logger.trace("Sending a small and large PingMessage to {}", to);
+        IAsyncCallback latencyProbeHandler = new IAsyncCallback()
```

Comment:
You can create two settable futures here (or the completable equivalent) and separate callbacks for small and large that point to the settable futures. You can then create a composite future and add a listener to it that does the right thing with the resulting largest/worst latency value.

I think this means you still need the enum, but we don't actually need to change the API of callbacks since we will never report a probe via a callback directly.

Reply:
Hm, ok I'm not entirely sure I follow but I will take a whack at it and get back to you!

Reply:
In the latest version I kept the changed API but I can remove it you think it'll make the PR easier to review/merge.